providers/gandi.go

```diff
@@ -0,0 +1,278 @@
+package providers
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/rancher/external-dns/dns"
+	gandi "github.com/prasmussen/gandi-api/client"
+	gandiDomain "github.com/prasmussen/gandi-api/domain"
+	gandiZone "github.com/prasmussen/gandi-api/domain/zone"
+	gandiZoneVersion "github.com/prasmussen/gandi-api/domain/zone/version"
+	gandiRecord "github.com/prasmussen/gandi-api/domain/zone/record"
+)
+
+type GandiHandler struct {
+	record      *gandiRecord.Record
+	zone        *gandiZone.ZoneInfoBase
+	zoneVersion *gandiZoneVersion.Version
+	root        string
+	zoneDomain  string
+	zoneSuffix  string
+	sub         string
+}
+
+func init() {
+	gandiHandler := &GandiHandler{}
+
+	apiKey := os.Getenv("GANDI_APIKEY")
+	if len(apiKey) == 0 {
+		logrus.Infof("GANDI_APIKEY is not set, skipping init of %s provider", gandiHandler.GetName())
+		return
+	}
+
+	systemType := gandi.Production
+	testing := os.Getenv("GANDI_TESTING")
+	if len(testing) != 0 {
+		logrus.Infof("GANDI_TESTING is set, using testing platform")
+		systemType = gandi.Testing
+	}
+
+	if err := RegisterProvider("gandi", gandiHandler); err != nil {
```

Comment:
looks like line 43-45 can be moved to the line 35. Sp if the provider fails to register, there is no need to process GANDI_TESTING env var